### PR TITLE
[Workflow] Workflow tail recursion optimization

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

When a nested workflow step returns a sub-workflow, usually they execute in different workers. However this is not necessary when both workflow steps requires exactly same resources. In this case we can execute the subworkflow in-place. This is known as kinds of tail recursion optimization in some systems.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/workflow/step_executor.py [hunk 1]
```diff
@@ -372,7 +387,13 @@ def _workflow_step_executor(
             # Execute sub-workflow. Pass down "outer_most_step_id".
             with workflow_context.fork_workflow_step_context(
                     outer_most_step_id=outer_most_step_id):
-                result = execute_workflow(persisted_output)
+                sub_workflow = persisted_output
+                if is_same_ray_options(
```
Comment (stephanie-wang) [inline python/ray/workflow/step_executor.py:391]: Can you add a comment to explain why we execute in-place here?

File: python/ray/workflow/utils.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    accelerator_type = ray_options.get("accelerator_type")
+
+    resources = resources_from_resource_arguments(
+        num_cpus, num_gpus, memory, object_store_memory, _resources,
+        accelerator_type, num_cpus, num_gpus, memory, object_store_memory,
+        _resources, accelerator_type)
+    return resources
+
+
+def is_same_ray_options(ray_options_1: Dict, ray_options_2: Dict) -> bool:
```
Comment (stephanie-wang) [inline python/ray/workflow/utils.py:20]:
```suggestion
def resources_equal(ray_options_1: Dict, ray_options_2: Dict) -> bool:
```

**Review by stephanie-wang — COMMENTED** [review]
The implementation and tests look good. Could you just add a PR description and some comments to explain why we do this?

**Review by stephanie-wang — APPROVED** [review]

**Comment by suquark** [comment]
There are some bugs about metadata. I'm fixing it.

**Comment by suquark** [comment]
Turns out the cause is a possible buggy step status update in the upstream codebase. I create a workaround and mark it with TODOs.

**Comment by suquark** [comment]
Massive unrelated errors occur in the CI. I'll try squash and rebase on the upstream.

File: python/ray/workflow/step_executor.py [hunk 3]
```diff
@@ -137,7 +138,8 @@ def _resolve_step_inputs(
     return signature.recover_args(flattened_args)
 
 
-def execute_workflow(workflow: "Workflow") -> "WorkflowExecutionResult":
+def execute_workflow(workflow: "Workflow",
+                     inplace=False) -> "WorkflowExecutionResult":
```
Comment (iycheng) [inline python/ray/workflow/step_executor.py:142]: Please update the comment about the arguments here.

File: python/ray/workflow/step_executor.py [hunk 4]
```diff
@@ -149,18 +151,42 @@ def execute_workflow(workflow: "Workflow") -> "WorkflowExecutionResult":
...
     step_options = workflow_data.step_options
-    persisted_output, volatile_output = _workflow_step_executor.options(
-        **step_options.ray_options).remote(
+    if inplace:
+        # TODO(suquark): For inplace execution, it is impossible
+        # to get the ObjectRef of the output before execution.
+        # Here we use a dummy ObjectRef, because _record_step_status does not
+        # even use it (?!).
+        _record_step_status(workflow.step_id, WorkflowStatus.RUNNING,
+                            [ray.put(None)])
```
Comment (iycheng) [inline python/ray/workflow/step_executor.py:160]: I feel this looks hacky. I feel there might be a bug in the current code, and it should be stored in step cache.

File: python/ray/workflow/step_executor.py [hunk 5]
```diff
@@ -372,7 +397,18 @@ def _workflow_step_executor(
             # Execute sub-workflow. Pass down "outer_most_step_id".
             with workflow_context.fork_workflow_step_context(
                     outer_most_step_id=outer_most_step_id):
-                result = execute_workflow(persisted_output)
+                sub_workflow = persisted_output
+                if resources_equal(runtime_options.ray_options,
```
Comment (iycheng) [inline python/ray/workflow/step_executor.py:401]: I'm not sure it's valid here. At least for memory usage, it's dynamic and might not be qualified to run.

For example, first one, run and put 10gb into memory and the second one might not be ok to run.

File: python/ray/workflow/step_executor.py [hunk 6]
```diff
@@ -372,7 +397,18 @@ def _workflow_step_executor(
...
+                                   sub_workflow.data.step_options.ray_options):
+                    # The workflow step returned by the current step has
+                    # exactly same resource and environment constraints.
+                    # So we execute it inplace.
+                    # TODO(suquark): We still have recursive Python calls.
+                    # This would cause stack overflow if we have a really
+                    # deep recursive call. We should fix it later.
```
Comment (iycheng) [inline python/ray/workflow/step_executor.py:408]: This looks dangerous. Should we have a feature flag for this feature? For example, .options(allow_inplace=True) then it'll run with in place optimization.

File: python/ray/workflow/step_executor.py [hunk 7]
```diff
@@ -149,18 +151,42 @@ def execute_workflow(workflow: "Workflow") -> "WorkflowExecutionResult":
...
+        if not isinstance(persisted_output, ray.ObjectRef):
+            volatile_output = ray.put(volatile_output)
+    else:
+        persisted_output, volatile_output = (
+            _workflow_step_executor_remote.options(
+                **step_options.ray_options).remote(
+                    workflow_data.func_body,
+                    workflow_context.get_workflow_step_context(),
+                    workflow.step_id, baked_inputs,
+                    workflow_data.step_options))
```
Comment (iycheng) [inline python/ray/workflow/step_executor.py:178]: Reorganize this a little bit:
```
executor_step_func = _step_output_cache if in_place else _workflow_step_executor_remote.options(
                **step_options.ray_options).remote
executor_step_func(...)
```

**Review by fishbone — CHANGES_REQUESTED** [review]
Could we have an option inside run, and allow user to specify in place or not?

File: python/ray/workflow/step_executor.py (hunk 4 above)
Comment (suquark) [reply python/ray/workflow/step_executor.py:160]: In the current code, `_record_step_status` and later calls never uses the third argument. This is very confusing.

Comment (iycheng) [reply python/ray/workflow/step_executor.py:160]: Agreed. I think there is a bug about this one. Let me check it.

I'm thinking about potential way to do this, but feel ok for now if we have a feature flag in the run.

One way might be, if we can make sure remote task is running in the same node, we probably can use local lock for it. Or we can use db's feature for this, for example, wait until something is there and load it.

File: doc/source/workflows/advanced.rst [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+    def exp_inplace(k, n, worker_id=None):
+        if n == 0:
+            return k
+        return exp_inplace.options(allow_inplace=True).step(
+            2 * k, n - 1, worker_id)
+
+With ``allow_inplace=True``, the step executes in the function that called ``.step()``. Ray options are ignored because they are used for remote execution.
+
+Inplace is also useful when you need to pass something that is only valid in the current process/physical machine to another step. For example:
```
Comment (iycheng) [inline doc/source/workflows/advanced.rst:33]: also mention that for inplace steps, step results won't be able to retrieve before the step is finished.

**Review by fishbone — APPROVED** [review]
It looks good. Please address the comment before merging.

**Comment by suquark** [comment]
The CI error seems unrelated. I'll merge the PR.

**Pull request merged**